src/routes/coingeckoProxy/index.ts

```diff
@@ -0,0 +1,93 @@
+import fastifyCaching from "@fastify/caching";
+import httpProxy from "@fastify/http-proxy";
+import { FastifyPluginAsync } from "fastify";
+import { ReadableStream } from "stream/web";
+
+const CACHE_TTL = 5 * 60 * 1000; // 5 min in ms
+
+type CachedData = {
+  item: {
+    contents: string;
+    headers: Headers;
+  };
+};
+
+function isCachedResponse(cached: any): cached is CachedData {
+  return (
+    cached &&
+    typeof cached === "object" &&
+    "item" in cached &&
+    typeof cached.item === "object" &&
+    "contents" in cached.item
+  );
+}
+
+const coingeckoProxy: FastifyPluginAsync = async (
+  fastify,
+  opts
+): Promise<void> => {
+  // Set up caching
+  fastify.register(fastifyCaching, { expiresIn: 300, serverExpiresIn: 300 });
+
+  // Intercepts request and uses cache if found
+  fastify.addHook("onRequest", async function (req, reply) {
+    fastify.cache.get(req.url.toLowerCase(), (err, result) => {
+      if (err) {
+        fastify.log.warn(`Error fetching cache for '${req.url}'`, err);
+        return err;
+      }
+      if (isCachedResponse(result)) {
+        fastify.log.info(`onRequest: cache hit for ${req.url}`);
+        reply
+          .headers({ "cache-hit": true, ...result.item.headers })
```

Comment:
Set `cache-hit` header for internal flow control and cached headers part of the original response.